Engine API: extend semantics of `executePayload` and `forkchoiceUpdated` methods (#165)
ethereum/execution-apis: src/engine/specification.md

1. Client software **MAY** provide additional details on the validation error if a payload is deemed `INVALID` by assigning the corresponding message to the `validationError` field.

1. The process of validating a payload on the canonical chain **MUST NOT** be affected by an active sync process on a side branch of the block tree. For example, if side branch `B` is `SYNCING` but the requisite data for validating a payload from canonical branch `A` is available, client software **MUST** run full validation of the payload and respond accordingly.

### Sync

In the context of this specification, the sync is understood as the process of obtaining data required to validate a payload. The sync process may consist of the following stages:

1. Pulling data from remote peers in the network.
1. Passing ancestors of a payload through the [Payload validation](#payload-validation) and obtaining a parent state.

> *Note:* Each of these stages is optional. Exact behavior of a client software during the sync process is implementation dependent.

Reviewer: ```suggestion
*Note:* Each of these stages is optional. Exact behavior of client software during the sync process is implementation dependent.
```

1. Client software **SHOULD** build the initial version of the payload which has an empty transaction set.

1. Client software **SHOULD** start the process of updating the payload. The strategy of this process is implementation dependent. The default strategy is to keep the transaction set up-to-date with the state of local mempool.

1. Client software **SHOULD** stop the updating process when either a call to `engine_getPayload` with the build process's `payloadId` is made or [`SECONDS_PER_SLOT`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#time-parameters-1) (12s in the Mainnet configuration) have passed since the point in time identified by the `timestamp` parameter.

## Core

### engine_newPayloadV1

#### Request

* method: `engine_newPayloadV1`
* params: 
  1. [`ExecutionPayloadV1`](#ExecutionPayloadV1)

#### Response

* result: [`PayloadStatusV1`](#PayloadStatusV1)
* error: code and message set in case an exception happens while processing the payload.

#### Specification

> 1. Client software **MUST** validate `blockHash` value as being equivalent to `Keccak256(RLP(ExecutionBlockHeader))`, where `ExecutionBlockHeader` is an object having a PoW block header structure. Fields of this object are set to the corresponding payload values and constant values according to the Block structure section of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#block-structure), extended with the corresponding section of [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399#block-structure). Client software **MUST** run this validation in all cases even if this branch or any other branches of the block tree are in an active sync process.

Reviewer: ```suggestion
1. Client software **MUST** validate `blockHash` value as being equivalent to `Keccak256(RLP(ExecutionBlockHeader))`, where `ExecutionBlockHeader` is the execution-layer block header (the former PoW block header structure). Fields of this object are set to the corresponding payload values and constant values according to the Block structure section of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#block-structure), extended with the corresponding section of [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399#block-structure). Client software **MUST** run this validation in all cases even if this branch or any other branches of the block tree are in an active sync process.
```

1. Client software **MAY NOT** validate the payload if the payload doesn't belong to the canonical chain.

1. Client software **MUST** respond to this method call in the following way:
  * `{status: INVALID_BLOCK_HASH, latestValidHash: null, validationError: null}` if the `blockHash` validation has failed
  * `{status: SYNCING, latestValidHash: null, validationError: null}` if the payload extends the canonical chain and requisite data for its validation is missing
  * with the payload status obtained from the [Payload validation](#payload-validation) process if the payload has been fully validated while processing the call
  * `{status: ACCEPTED, latestValidHash: null, validationError: null}` if the following conditions are met:
    - the `blockHash` of the payload is valid
    - the payload doesn't extend the canonical chain
    - the payload hasn't been fully validated
  * with `-32002: Invalid terminal block` error if terminal block conditions aren't met.

> 1. If any of the above fails due to errors unrelated to the normal processing flow of the method, a client software **MUST** respond with an error object.

Reviewer: ```suggestion
1. If any of the above fails due to errors unrelated to the normal processing flow of the method, client software **MUST** respond with an error object.
```

* error: code and message set in case an exception happens while processing the payload.

#### Specification

1. Client software **MUST** validate `blockHash` value as being equivalent to `Keccak256(RLP(ExecutionBlockHeader))`, where `ExecutionBlockHeader` object is composed of the payload field values and constant values described in the [Block structure](https://eips.ethereum.org/EIPS/eip-3675#block-structure) section of the [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification). Additionally:
    * Client software **MUST** return `{status: INVALID_BLOCK_HASH, latestValidHash: null, validationError: null}` and discard the payload if this validation fails
    * Client software **MUST** run this validation in all cases even if this branch or any other branches of the block tree are in an active sync process.

1. Client software **SHOULD** initiate the sync process if requisite data for payload validation is missing. The sync process is specified in the [Sync process](#sync-process) section. Additionally:
    * Client software **MUST** return `{status: SYNCING, latestValidHash: null, validationError: null}` if the sync process has been initiated.

> 1. Client software **SHOULD** validate the payload if requisite data for payload validation is locally available. The validation process is specified in the [Payload validation process](#payload-validation-process) section. Additionally:

Reviewer 2: From acd today: this should be made stricter, to not make it implementation-dependent so that clients all lazily choose _not_ to validate any `newPayload`s at all

    * Client software **MUST** return the result of the payload validation if the validation has been initiated.

1. Client software **MUST** return `{status: ACCEPTED, latestValidHash: null, validationError: null}` if neither payload validation nor sync process has been initiated.

1. If any of the above fails due to errors unrelated to the normal processing flow of the method, the client software **MUST** return an error.

> ##### Payload validation process

Author: I'd also move `##### Payload build process` section to the `## Routines`, thoughts?

> ##### Payload validation process

Reviewer: Yep, makes sense

Payload validation process consists of validating a payload with respect to the block header and execution environment rule sets. The process is specified as follows:

1. Client software **MUST** validate the payload according to the block header and execution environment rule set with modifications to these rule sets defined in the [Block Validity](https://eips.ethereum.org/EIPS/eip-3675#block-validity) section of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification). Additionally:
    * If validation succeeds, the response **MUST** contain `{status: VALID, latestValidHash: payload.blockHash}`
    * If validation fails, the response **MUST** contain `{status: INVALID, latestValidHash: validHash}` where `validHash` is the block hash of the most recent *valid* ancestor of the invalid payload. That is, the valid ancestor of the payload with the highest `blockNumber`

>     * Client software **MUST** discard the payload if it's deemed invalid.

Author: It may be the case but it is not in any EL client implementations today, AFAIK. We can say it in a different way:
> Client software **MUST NOT** serve the payload if it's deemed invalid, including serving over the wire and via any API endpoint.

> 1. Client software **MUST** return `-32002: Invalid terminal block` error if the parent block is locally available and is a PoW block that doesn't satisfy terminal block conditions according to [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#definitions). This check maps on the Transition block validity section of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#transition-block-validity).

Author: I was thinking about this edge case. It seems like a client software can't respond with error because it cannot validate `A` with respect to the terminal block conditions if `A` is missing. In this case it should rather respond with `INVALID`, but then the question what should it return as `latestValidHash`. It should be `latestValidHash: parentOfA.block_hash`, but CL won't be able to do anything with this value as there is no such block in its database. Probably, CL may look at the `parent_hash` of `A` in this case to conclude that a blockchain starting at `A` and ending at `B` is invalid.

I would replace this error with `INVALID` but need to decide is returned as `latestValidHash`.

1. Client software **MAY** provide additional details on the validation error if the payload is deemed `INVALID` by assigning the corresponding message to the `validationError` field.

1. The process of validating a payload on the canonical chain **MUST NOT** be affected by an active sync process on a side branch of the block tree. For example, if side branch `B` is `SYNCING` but the requisite data for validating a payload from canonical branch `A` is available, client software **MUST** initiate the validation process.

> ##### Sync process

Reviewer: Same with this. Let's put it under `## Routines` or some more generic header for reference

#### Response

* result: `object`

>     - `status`: `enum` - `"VALID" | "INVALID" | "SYNCING" | "ACCEPTED"`

Reviewer: Ditto on from `newPayload`. I suggest returning a `payloadStatus` object along with `payloadId` for better encapsulation. And specific which `status`s can be returnd here from a global set (i.e. the full set less `INVALID_BLOCK_HASH`)

    - `latestValidHash`: `DATA|null`, 32 Bytes - the hash of the most recent *valid* block in the branch defined by payload and its ancestors
    - `validationError`: `String|null` - a message providing additional details on the validation error if the payload is deemed `INVALID`
    - `payloadId`: `DATA|null`, 8 Bytes - identifier of the payload build process or `null`

> * error: code and message set in case an exception happens while validating payload, updating the forkchoice or initiating the payload build process.

Reviewer: ```suggestion
* error: code and message set in case an exception happens while the validating payload, updating the forkchoice or initiating the payload build process.
```

#### Specification

1. Client software **SHOULD** initiate the sync process if `forkchoiceState.headBlockHash` references an unknown payload or a payload that can't be validated because requisite data is missing. The sync process is specified in the [Sync process](#sync-process) section. Additionally:

>     * Client software **MUST** return `{status: SYNCING, latestValidHash: null, validationError: null, payloadId: null}` if the sync process has been initiated.

Reviewer: If the client doesn' follow the SHOULD here, MUST they still return SYNCING?

Reviewer: `newBlock` says the following:

> Client software **SHOULD** initiate the sync process if requisite data for payload validation is missing. The sync process is specified in the [Sync process](#sync-process) section. Additionally:

The only difference is that it doesn't mentioned `headBlockHash` specifically. Should these actually be the same text? *And* if so, I would recommend hoisting this call into the `validationProcess`. `payloadId` can conditionally be returned as `null` if `SYNCING` toward the end of the list (outside `validationProcess`)

Author: > If the client doesn' follow the SHOULD here, MUST they still return SYNCING?

Yes, it must if we want to keep this semantics unchanged with respect to the previous version.

> And if so, I would recommend hoisting this call into the validationProcess.

Good suggestion. I was leaving it here only for explicitly specifying return values for each branch of method processing. But I think we can improve on return values as well by moving them in a separate section for this method, considering this move for `newPayload` as well -- will see how it applies.

> 1. Client software **MAY** skip an update of the forkchoice state and **MAY NOT** begin a payload build process if the payload identified by `forkchoiceState.headBlockHash` hasn't been validated yet. Additionally:

Reviewer: If you follow the above suggestion to insert `SYNCING` logic into `payloadValidation`, then you need to update this statement to say:

> if the payload identified by `forkchoiceState.headBlockHash` known but has not been validated yet

    * Client software **MUST** return `{status: ACCEPTED, latestValidHash: null, validationError: null, payloadId: null}` in this case.

> 1. Client software **MAY** skip an update of the forkchoice state and **MAY NOT** begin a payload build process if `forkchoiceState.headBlockHash` references an ancestor of the head of the canonical chain. Additionally:

Reviewer: Should this be a **MUST**? and **MUST NOT**?

Otherwise, it leaves open the option to use this API to (if supported by EL) rewind the fork choice and also build old payloads. The former (rewinding) might *only* be useful to user APIs but seems an abuse of this API. And the latter (building old payloads) might only really be used for attacks and other nuisance messages.

Author: Another option is to **MAY** update and **MUST NOT** begin build process. I think that restricting rewinding the head is a bit awkward.

Reviewer: okay, that seems generally fine but see my comment about block leaves below

    * Client software **MUST** return `{status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null, payloadId: null}` in this case.

1. Before updating the forkchoice state, client software **MUST** ensure the validity of the payload identified by `forkchoiceState.headBlockHash`, and **MAY** initiate payload validation process if that need be. The validation process is specified in the [Payload validation process](#payload-validation-process) section. Additionally:

>     * Client software **MUST NOT** update the forkchoice state and immediately respond to the method call with the corresponding result if the payload or any of its ancestors is deemed `INVALID`.

Reviewer: ```suggestion
    * Client software **MUST NOT** update the forkchoice state and immediately respond to this call with the corresponding result if the payload or any of its ancestors is deemed `INVALID`.
```

> 1. Client software **MUST** update its forkchoice state if payloads identified by `forkchoiceState.headBlockHash` and `forkchoiceState.finalizedBlockHash` are deemed valid. The update is specified as follows:

Reviewer: ```suggestion
1. Client software **MUST** update its forkchoice state if payloads identified by `forkchoiceState.headBlockHash` and `forkchoiceState.finalizedBlockHash` are `VALID`. The update is specified as follows:
```

    * The values `(forkchoiceState.headBlockHash, forkchoiceState.finalizedBlockHash)` of this method call map on the `POS_FORKCHOICE_UPDATED` event of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification) and **MUST** be processed according to the specification defined in the EIP
    * All updates to the forkchoice state resulting from this call **MUST** be made atomically

>     * Client software **MUST** return `{status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null, payloadId: null}` if `payloadAttributes` is `null`.

Reviewer: I suggest not touching `payloadId` until all of the sync/validation stuff is done.

Then get to `payloadId` cleany below -- e.g.
* "If INVALID or SYNCING (no forkchoice update), don't initiate build process, and return `payloadId` as `null`."
* "if VALID and `payloadAttributes` is null, `payloadId` is null"
* "If VALID and `payloadAttributes` is not null, initiate build process and return `payloadId`

1. Client software **MUST** respond to this method call in the following way:
  * `{payloadStatus: {status: SYNCING, latestValidHash: null, validationError: null}, payloadId: null}` if `forkchoiceState.headBlockHash` references an unknown payload or a payload that can't be validated because requisite data is missing
  * `{payloadStatus: {status: INVALID, latestValidHash: null, validationError: errorMessage | null}, payloadId: null}` obtained from the [Payload validation](#payload-validation) process if the payload is deemed `INVALID`
  * `{payloadStatus: {status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null}, payloadId: null}` if the payload is deemed `VALID` and the build process hasn't been started
  * `{payloadStatus: {status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null}, payloadId: buildProcessId}` if the payload is deemed `VALID` and the build process has begun.

> 1. If any of the above fails due to errors unrelated to the normal processing flow of the method, the client software **MUST** return an error.

Reviewer 3: So errors in the payload assembling process are returned as JSON_RPC errors directly?

Author: Not sure I understand your question correctly. But if an exception occurred during initiating a payload build process then the call of this method must be responded with error.

Generally, this point states that in the case when e.g. a payload is `INVALID` according to EE rules an `INVALID` status must be in the response, but if exception happened during the validation then the call must be responded with error, not an `INVALID` status.

Reviewer 4: What should EL return if we receive the wrong timestamp in payload attributes? Because I think `INVALID` is connected to headBlockHash, not payload attributes. By the wrong timestamp, I mean `headBlockHash.Timestamp >= payloadAttributes.Timestamp`

Author: It would mean that CL is faulty, EL may respond with error in this case. There is `-32602: Invalid params` error that fits well into this case

#### Response

* result: `object`
  - `payloadStatus`: [`PayloadStatusV1`](#PayloadStatusV1); values of the `status` field in the context of this method are restricted to the following subset:
    * `"VALID"`
    * `"INVALID"`
    * `"SYNCING"`
  - `payloadId`: `DATA|null`, 8 Bytes - identifier of the payload build process or `null`
* error: code and message set in case an exception happens while the validating payload, updating the forkchoice or initiating the payload build process.

#### Specification

1. Client software **MAY** initiate the sync process if `forkchoiceState.headBlockHash` references an unknown payload or a payload that can't be validated because requisite data is missing. The sync process is specified in the [Sync](#sync) section.

> 1. Client software **MAY** skip an update of the forkchoice state and **MUST NOT** begin a payload build process if `forkchoiceState.headBlockHash` references an ancestor of the head of the canonical chain.

Reviewer: Should this be any non-leaf value? -- That is you MAY skip it if it isn't a leaf of the block tree

Otherwise you can abuse the rewind (turn this MAY into a MUST) by fcu to a non-canonical branch and then fcu to a non-leaf block in what was previously the canonical chain

#### Response

* result: `object`
  - `payloadStatus`: [`PayloadStatusV1`](#PayloadStatusV1); values of the `status` field in the context of this method are restricted to the following subset:
    * `"VALID"`
    * `"INVALID"`
    * `"SYNCING"`
  - `payloadId`: `DATA|null`, 8 Bytes - identifier of the payload build process or `null`
* error: code and message set in case an exception happens while the validating payload, updating the forkchoice or initiating the payload build process.

#### Specification

1. Client software **MAY** initiate a sync process if `forkchoiceState.headBlockHash` references an unknown payload or a payload that can't be validated because data that are requisite for the validation is missing. The sync process is specified in the [Sync](#sync) section.

1. Client software **MAY** skip an update of the forkchoice state and **MUST NOT** begin a payload build process if `forkchoiceState.headBlockHash` doesn't reference a leaf of the block tree. That is, the block referenced by `forkchoiceState.headBlockHash` is neither the head of the canonical chain nor a block at the tip of any other chain.

> 1. Client software **MUST** return `-32002: Invalid terminal block` error if `forkchoiceState.headBlockHash` references a PoW block that doesn't satisfy terminal block conditions according to [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#definitions). This check maps on the Transition block validity section of the EIP.

Reviewer: ```suggestion
1. Client software **MUST** return `-32002: Invalid terminal block` error if `forkchoiceState.headBlockHash` references a PoW block that doesn't satisfy terminal block conditions according to [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#definitions). This check maps to the transition block validity section of the EIP.
```

1. Client software **MUST** update its forkchoice state if payloads referenced by `forkchoiceState.headBlockHash` and `forkchoiceState.finalizedBlockHash` are `VALID`. The update is specified as follows:
  * The values `(forkchoiceState.headBlockHash, forkchoiceState.finalizedBlockHash)` of this method call map on the `POS_FORKCHOICE_UPDATED` event of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#block-validity) and **MUST** be processed according to the specification defined in the EIP
  * All updates to the forkchoice state resulting from this call **MUST** be made atomically.

1. Client software **MUST** begin a payload build process building on top of `forkchoiceState.headBlockHash` and identified via `buildProcessId` value if `payloadAttributes` is not `null` and the forkchoice state has been updated successfully. The build process is specified in the [Payload building](#payload-building) section.

1. Client software **MUST** respond to this method call in the following way:

>   * `{payloadStatus: {status: SYNCING, latestValidHash: null, validationError: null}, payloadId: null}` if `forkchoiceState.headBlockHash` references an unknown payload or a payload that can't be validated because data that are requisite for the validation is missing

Reviewer: ```suggestion
  * `{payloadStatus: {status: SYNCING, latestValidHash: null, validationError: null}, payloadId: null}` if `forkchoiceState.headBlockHash` references an unknown payload or a payload that can't be validated because requisite data for the validation is missing
```

  * `{payloadStatus: {status: INVALID, latestValidHash: null, validationError: errorMessage | null}, payloadId: null}` obtained from the [Payload validation](#payload-validation) process if the payload is deemed `INVALID`
  * `{payloadStatus: {status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null}, payloadId: null}` if the payload is deemed `VALID` and a build process hasn't been started
  * `{payloadStatus: {status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null}, payloadId: buildProcessId}` if the payload is deemed `VALID` and the build process has begun

>   * with `-32002: Invalid terminal block` error if terminal block conditions aren't met.

Reviewer: ```suggestion
  * with `-32002: Invalid terminal block` error if terminal block conditions aren't met
```
